SQL: Add CCS SQL documentation

bpintea opened the pull request:
This adds the documentation for CCS SQL.

A few other edits have also been applied, around syntax definition using
 `[]` and `?`, as well as capitalisation.

It also introduces a new discriminator for the CSV tests, a suffix,
`-NoEdit`, which prevents the testing code from altering a query to turn
it into a CCS query. This is meant for those queries written as CCS
queries already (those to be used as-is for the docs).

bpintea committed:
Add CCS SQL documentation

This adds the documentation for CCS SQL.

It also introduces a new discriminator for the CSV tests, a prefix,
`-NoEdit`, which prevents the testing code from altering a query to turn
it into a CCS query. This is meant for those queries written as CCS
queries already.

bpintea committed:
Fix file reference

Correct referenced file name.

bpintea committed:
Further doc fixes

- remove superfluous `?` after SHOW FUNCTIONS pattern;
- add "show catalogs" page to index.
- bridge works in "catalog_identifier".

bpintea committed:
Merge branch 'master' into doc/sql_ccs

bpintea committed:
Include missing resource

Include new show-catalogs.asciidoc resource.

bpintea committed:
Merge branch 'master' into doc/sql_ccs

elasticmachine commented:
Pinging @elastic/es-docs (Team:Docs)

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

Luegg commented on x-pack/plugin/sql/qa/server/multi-cluster-with-security/src/test/java/org/elasticsearch/xpack/sql/qa/multi_cluster_with_security/JdbcCsvSpecIT.java:
```diff
@@ -94,7 +101,9 @@ private static CsvTestCase qualifyFromClause(CsvTestCase testCase) {
-        if (FROM_QUALIFIED.matcher(csvTestCase().query).matches() == false) {
+        // Only set the default catalog if the query index isn't yet qualified with the catalog, which can happen if query has been written
+        // qualified from the start (for the documentation) or edited in qualifyFromClause() above.
+        if (mayEditQuery(testName) && FROM_QUALIFIED.matcher(csvTestCase().query).matches() == false) {
```
Instead of having the `-NoEdit` suffix, wouldn't it be possible to change the `FROM_QUALIFIED` pattern to match any qualified index name? Shouldn't anything with a colon in the index name be kept from rewriting?

costin replied:
👍 
Determine whether to add or replace the catalog based on the presence or lack of the catalog in the name.

bpintea replied:
> wouldn't it be possible to change the FROM_QUALIFIED pattern to match any qualified index name?

For this one case, sure, it could be done quite unequivocally. But generally the qualification can be done as:
* `FROM "remote*cluster":foo`
* `FROM remote_cluster:foo`
* `FROM "remote_cluster:foo"`
* `FROM "remote*cluster":"foo"`

The `FROM_QUALIFIED` is safe enough, since the query editing is done in the same class, but generally I thought that writing a regex that'll match whatever documentation queries we might want to add and also be reliable enough (`SELECT SUBSTRING("SELECT foo FROM remote_index:foo") ...`) might not be the best solution.

> Determine whether to add or replace the catalog based on the presence or lack of the catalog in the name.

The catalog is not used in the doc query, it's a wildcard (to be more suggestive).

Anyways, I'll add regexp matching and we can change it if ever needed.

bpintea replied:
Fixed to check the FROM clause with a regex. It's fragile, but should do for now.

jrodewig reviewed (commented):
Thanks @bpintea. This looks great overall, but we may also want to update a few additional pages:

* Add the `catalog` request body parameter to the [SQL search API reference docs](https://www.elastic.co/guide/en/elasticsearch/reference/master/sql-search-api.html#sql-search-api-request-body)
* Mention the `<remote_cluster>:<target>` syntax on the [SQL index patterns page](https://www.elastic.co/guide/en/elasticsearch/reference/master/sql-index-patterns.html#sql-index-patterns-multi).
* Add the SQL search API to the [Supported CCS APIs list](https://www.elastic.co/guide/en/elasticsearch/reference/7.16/modules-cross-cluster-search.html#ccs-supported-apis).

I also left a few minor suggestions you can take or leave as wanted.

bpintea committed:
Apply suggestions from code review

Co-authored-by: James Rodewig <james.rodewig@elastic.co>

costin commented on x-pack/plugin/sql/qa/server/multi-cluster-with-security/src/test/java/org/elasticsearch/xpack/sql/qa/multi_cluster_with_security/JdbcCsvSpecIT.java:
```diff
@@ -27,6 +27,7 @@
     public static final String REMOTE_CLUSTER_NAME = "my_remote_cluster"; // gradle defined
     public static final String EXTRACT_FN_NAME = "EXTRACT";
 
+    private static final String NO_EDIT_SUFFIX = "-NoEdit";
```
This flag doesn't have a great name and is redundant. It's an indicator that the ccs processing is too intrusive.
How about looking for qualified vs non-qualified names instead to differentiate between where to add the catalog or not?

bpintea committed:
Address review comments

- Add the catalog request body parameter to the SQL search API reference
  docs
- Mention the <remote_cluster>:<target> syntax on the SQL index patterns
page.
- Add the SQL search API to the Supported CCS APIs list.

- various rephrasings.

costin commented on x-pack/plugin/sql/qa/server/src/main/resources/multi-cluster-with-security/multi-cluster-docs.csv-spec:
```diff
@@ -0,0 +1,17 @@
+
+fromQualifiedTableQuoted-NoEdit
+// tag::fromQualifiedTableQuoted
+SELECT emp_no FROM "my*cluster:*emp" LIMIT 1;
```
already qualified means `-NoEdit`.

costin reviewed (commented):
Left some comments, LGTM otherwise.

bpintea committed:
Merge branch 'master' into doc/sql_ccs

bpintea commented on docs/reference/sql/endpoints/jdbc.asciidoc:
```diff
@@ -148,6 +148,10 @@ will be - typically the first in natural ascending order) for fields with multip
 
+[discrete]
+==== Cluster
+`catalog`:: The default catalog to execute the query under.
```
Thanks, @jrodewig. I did rephrase it, but not exactly as suggested (as there is no defaulting per se). But happy to apply further improvements.

bpintea commented on docs/reference/sql/language/syntax/commands/select.asciidoc:
```diff
@@ -123,6 +123,12 @@ The name can be a <<multi-index, pattern>> pointing to multiple indices (likely
 include-tagged::{sql-specs}/docs/docs.csv-spec[fromTablePatternQuoted]
 ----
 
+The index name or expression can also be qualified with a cluster name or wildcard, resulting in a <<modules-cross-cluster-search, cross cluster search>>.
```
Same here, rephrased, but slightly differently (since `<target>` will take a pattern, not just a wildcard)

bpintea committed:
Update docs/reference/sql/language/syntax/commands/describe-table.asciidoc

Co-authored-by: James Rodewig <james.rodewig@elastic.co>

jrodewig commented on docs/reference/search/search-your-data/search-across-clusters.asciidoc:
```diff
@@ -20,6 +20,7 @@ The following APIs support {ccs}:
 * <<multi-search-template,Multi search template>>
 * <<search-field-caps,Field capabilities>>
 * experimental:[] <<eql-search-api,EQL search>>
+* experimental:[] <<sql-search-api,SQL search>>
```
Is this experimental for SQL? If so, we should include this `experimental:[]` elsewhere in the docs.

bpintea replied:
@jrodewig I've added the `experimental:[]` tag elsewhere too, let me know please if this is OK. Thanks!

jrodewig replied:
I double-checked these, and the tags look good overall. I did leave some suggestions related to `cross cluster search` references, which should be hyphenated.

bpintea replied:
Thanks, @jrodewig!

jrodewig commented on docs/reference/sql/apis/sql-search-api.asciidoc:
```diff
@@ -54,8 +54,13 @@ precedence.
+
 `columnar`::
-(Optional, Boolean) If `true`, returns results in a columnar format. Defaults to
+(Optional, boolean) If `true`, returns results in a columnar format. Defaults to
```
I think this is supposed to be capitalized unless we're referring to something like the `bool` query.

bpintea replied:
The other types (`string`, `integer`, but also `object`, `time values`, `array`)  are all lowercase. Shouldn't we capitalize them all or none?

bpintea replied:
I see. Thanks, reverted.

jrodewig reviewed (approved):
LGTM. Thanks!

astefan commented on docs/reference/sql/language/syntax/commands/show-tables.asciidoc:
```diff
@@ -6,14 +6,18 @@
+              LIKE pattern]]?      <2>
+    [INCLUDE FROZEN]?              <3>
+    [table_identifier |            <4>
+     LIKE pattern]                 <5>
```
`[table_identifier | LIKE pattern]?` ?

bpintea replied:
Good catch, thanks, fixed.

astefan reviewed (approved):
LGTM.
Left two comments.

bpintea committed:
Address review comments

- remove query name tag and logic based on it.
- add missing `?` in syntax documentation.

bpintea committed:
Merge branch 'master' into doc/sql_ccs

bpintea commented on docs/reference/sql/language/syntax/commands/show-columns.asciidoc:
```diff
@@ -5,13 +5,18 @@
+SHOW COLUMNS
+    [CATALOG identifier]? <1>
+    [INCLUDE FROZEN]?     <2>
+    [FROM | IN]
```
Grammar makes it mandatory, this was wrong in the initial docs.

bpintea committed:
Merge branch 'master' into doc/sql_ccs

bpintea committed:
Mark CCS SQL as experimental

Mark CCS SQL as experimental.

jrodewig commented on docs/reference/sql/endpoints/jdbc.asciidoc:
```diff
@@ -148,6 +148,13 @@ will be - typically the first in natural ascending order) for fields with multip
+`catalog`:: Default catalog (cluster) for queries. If unspecified, the
+queries execute on the data in the local cluster only.
++
+experimental:[] See <<modules-cross-cluster-search, cross cluster search>>.
```
```suggestion
experimental:[] See <<modules-cross-cluster-search,{ccs}>>.
```

It should be `cross-cluster search` with a hyphen. However, you can use the `{ccs}`, which auto-expands to that text.

jrodewig commented on docs/reference/sql/language/indices.asciidoc:
```diff
@@ -34,6 +34,19 @@ include-tagged::{sql-specs}/docs/docs.csv-spec[fromTablePatternQuoted]
 
 NOTE: There is the restriction that all resolved concrete tables have the exact same mapping.
 
+experimental:[] To run a <<modules-cross-cluster-search, cross cluster search>>, specify a
```
```suggestion
experimental:[] To run a <<modules-cross-cluster-search,{ccs}>>, specify a
```

jrodewig commented on docs/reference/sql/language/syntax/commands/select.asciidoc:
```diff
@@ -123,6 +123,17 @@ The name can be a <<multi-index, pattern>> pointing to multiple indices (likely
 include-tagged::{sql-specs}/docs/docs.csv-spec[fromTablePatternQuoted]
 ----
 
+experimental:[] To run a <<modules-cross-cluster-search, cross cluster search>>, specify a
```
```suggestion
experimental:[] To run a <<modules-cross-cluster-search,{ccs}>>, specify a
```

bpintea committed:
Apply suggestions from code review

Co-authored-by: James Rodewig <james.rodewig@elastic.co>

bpintea committed:
Apply comments suggestions

s/boolean/Boolean

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.0](https://github.com/elastic/elasticsearch/pull/82277)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/82277">](https://github.com/elastic/elasticsearch/pull/82277) |
| ❌ |  7.17  | Commit could not be cherrypicked due to conflicts |
| ❌ |  7.16  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 81545`
